File: tags/release-0.91/src/gui/parmbox3.cpp
// File:     parmbox3.cpp
// Location: src/gui
// Purpose:  implement ParmBox3 class
// Author:   Stephen Gold sgold@sonic.net
// (c) Copyright 2012 Stephen Gold
// Distributed under the terms of the GNU General Public License

/*
This file is part of the Gold Tile Game.

The Gold Tile Game is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by the 
Free Software Foundation, either version 3 of the License, or (at your 
option) any later version.

The Gold Tile Game is distributed in the hope that it will be useful, but 
WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY 
or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License 
for more details.

You should have received a copy of the GNU General Public License
along with the Gold Tile Game.  If not, see <http://www.gnu.org/licenses/>.
*/

#ifdef _WINDOWS
#include "gui/parmbox3.hpp"
#include "gui/resource.hpp"
#include "gui/win_types.hpp"

// message handler (callback) for this dialog
static int CALLBACK message_handler3(
    HWND windowHandle,
    MessageType message,
    WPARAM wParameter, 
    LPARAM lParameter)
{
    ASSERT(windowHandle != NULL);
    ParmBox3* const p_box = (ParmBox3*)BaseWindow::Lookup(windowHandle);
    ASSERT(HWND(*p_box) == windowHandle);
    int const result = p_box->HandleMessage(message, wParameter, lParameter);

    return result;
}


// lifecycle

ParmBox3::ParmBox3(GameOpt& rGameOpt):
Dialog("PARMBOX3", &message_handler3),
    mrGameOpt(rGameOpt)
{
}


// misc methods

/* static */ IdType ParmBox3::EditboxId(IdType sliderId) {
    IdType result = 0;

    switch (sliderId) {
    case IDC_SLIDER1:
        result = IDC_EDIT1;
        break;
    case IDC_SLIDER2:
        result = IDC_EDIT2;
        break;
    case IDC_SLIDER3:
        result = IDC_EDIT3;
        break;
    case IDC_SLIDER4:
        result = IDC_EDIT4;
        break;
    case IDC_SLIDER5:
        result = IDC_EDIT5;
        break;
    default:
        FAIL();
    }

    return result;
};

INT_PTR ParmBox3::HandleMessage(MessageType message, WPARAM wParam, LPARAM lParam) {
    INT_PTR result = FALSE;

    switch (message) {
    case WM_INITDIALOG: {
        Dialog::HandleMessage(message, wParam);

        InitControl(IDC_SLIDER1, mrGameOpt.HandsDealt(), 
            GameOpt::HANDS_DEALT_MIN, HANDS_DEALT_MAX);
        InitControl(IDC_SLIDER2, mrGameOpt.HandSize(), 
            GameOpt::HAND_SIZE_MIN, HAND_SIZE_MAX);
        InitControl(IDC_SLIDER3, mrGameOpt.AttrCnt(), 
            Combo::ATTR_CNT_MIN, Combo::ATTR_CNT_MAX);
        InitControl(IDC_SLIDER4, mrGameOpt.ClonesPerCombo(),
            0, CLONES_PER_COMBO_MAX);
        InitControl(IDC_SLIDER5, mrGameOpt.BonusPercent(),
            0, BONUS_PERCENT_MAX);

        result = TRUE;
        break;
                        }

    case WM_COMMAND: {
        IdType const id = LOWORD(wParam);
        switch (id) {
        case IDC_EDIT1:
        case IDC_EDIT2:
        case IDC_EDIT3:
        case IDC_EDIT4:
        case IDC_EDIT5: {
            ValueType const value = GetTextValue(id);
            if (value != VALUE_INVALID) {
                IdType const slider_id = SliderId(id);
                ValueType const slider_value = SetSliderValue(slider_id, value);
                if (slider_value != value) {
                    SetTextValue(id, slider_value);
                }
                UpdateValue(slider_id, slider_value);
            }
            break;
                        }
        }
        break;
                     }

    case WM_VSCROLL:
    case WM_HSCROLL: {
        IdType const slider_id = SliderId(HWND(lParam));
        ASSERT(slider_id != 0);
        ValueType const value = GetSliderValue(slider_id);
        IdType const editbox_id = EditboxId(slider_id);
        SetTextValue(editbox_id, value);
        UpdateValue(slider_id, value);
        break;
                     }
    }

    if (result == FALSE) {
        result = Dialog::HandleMessage(message, wParam);
    }

    return result;
}

void ParmBox3::InitControl(
    IdType sliderId,
    ValueType value,
    ValueType minValue,
    ValueType maxValue)
{
    ASSERT(value >= minValue);
    ASSERT(value <= maxValue);

    SetSliderRange(sliderId, minValue, maxValue);

    ValueType const slider_value = SetSliderValue(sliderId, value);
    ASSERT(slider_value == value);

    IdType const min_id = MinId(sliderId);
    SetTextValue(min_id, minValue);

    IdType const max_id = MaxId(sliderId);
    SetTextValue(max_id, maxValue);

    IdType const editbox_id = EditboxId(sliderId);
    SetTextValue(editbox_id, slider_value);
}

/* static */ IdType ParmBox3::MaxId(IdType sliderId) {
    IdType result = 0;

    switch (sliderId) {
    case IDC_SLIDER1:
        result = IDC_SMAX1;
        break;
    case IDC_SLIDER2:
        result = IDC_SMAX2;
        break;
    case IDC_SLIDER3:
        result = IDC_SMAX3;
        break;
    case IDC_SLIDER4:
        result = IDC_SMAX4;
        break;
    case IDC_SLIDER5:
        result = IDC_SMAX5;
        break;
    default:
        FAIL();
    }

    return result;
}

/* static */ IdType ParmBox3::MinId(IdType sliderId) {
    IdType result = 0;

    switch (sliderId) {
    case IDC_SLIDER1:
        result = IDC_SMIN1;
        break;
    case IDC_SLIDER2:
        result = IDC_SMIN2;
        break;
    case IDC_SLIDER3:
        result = IDC_SMIN3;
        break;
    case IDC_SLIDER4:
        result = IDC_SMIN4;
        break;
    case IDC_SLIDER5:
        result = IDC_SMIN5;
        break;
    default:
        FAIL();
    }

    return result;
}

/* static */ IdType ParmBox3::SliderId(IdType editboxId) {
    IdType result = 0;

    switch (editboxId) {
    case IDC_EDIT1:
        result = IDC_SLIDER1;
        break;
    case IDC_EDIT2:
        result = IDC_SLIDER2;
        break;
    case IDC_EDIT3:
        result = IDC_SLIDER3;
        break;
    case IDC_EDIT4:
        result = IDC_SLIDER4;
        break;
    case IDC_EDIT5:
        result = IDC_SLIDER5;
        break;
    default:
        FAIL();
    }

    return result;
}

IdType ParmBox3::SliderId(HWND handle) const {
    IdType result = 0;

    if (handle == GetControlHandle(IDC_SLIDER1)) {
        result = IDC_SLIDER1;
    } else if (handle == GetControlHandle(IDC_SLIDER2)) {
        result = IDC_SLIDER2;
    } else if (handle == GetControlHandle(IDC_SLIDER3)) {
        result = IDC_SLIDER3;
    } else if (handle == GetControlHandle(IDC_SLIDER4)) {
        result = IDC_SLIDER4;
    } else if (handle == GetControlHandle(IDC_SLIDER5)) {
        result = IDC_SLIDER5;
    } else {
        FAIL();
    }

    return result;
}

void ParmBox3::UpdateValue(IdType sliderId, ValueType value) {
    switch (sliderId) {
    case IDC_SLIDER1:
        mrGameOpt.SetHandsDealt(value);
        break;
    case IDC_SLIDER2:
        mrGameOpt.SetHandSize(value);
        break;
    case IDC_SLIDER3: {
        ASSERT(value >= Combo::ATTR_CNT_MIN);
        ASSERT(value <= Combo::ATTR_CNT_MAX);
        AttrCntType attr_cnt = AttrCntType(value);
        mrGameOpt.SetAttrCnt(attr_cnt);
        break;
                      }
    case IDC_SLIDER4:
        mrGameOpt.SetClonesPerCombo(value);
        break;
    case IDC_SLIDER5:
        mrGameOpt.SetBonusPercent(value);
        break;
    default:
        FAIL();
    }
}
#endif // defined(_WINDOWS)

File: trunk/Java/GoldTile/src/goldtile/ParmBox3.java
// File:     ParmBox3.java
// Location: Java/GoldTile/src/goldtile
// Purpose:  ParmBox3 class for the Gold Tile Game
/**
 * @author Stephen Gold
 */
// (c) Copyright 2012 Stephen Gold
// Distributed under the terms of the GNU General Public License

/*
This file is part of the Gold Tile Game.

The Gold Tile Game is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by the
Free Software Foundation, either version 3 of the License, or (at your
option) any later version.

The Gold Tile Game is distributed in the hope that it will be useful, but
WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
for more details.

You should have received a copy of the GNU General Public License
along with the Gold Tile Game.  If not, see <http://www.gnu.org/licenses/>.
*/

package goldtile;

import java.awt.Dimension;
import javax.swing.*;

public class ParmBox3
    extends WizardCard
    implements javax.swing.event.ChangeListener  // spinners
{
    // constants, sorted by type
    final private static Dimension SMALL = new Dimension(45, 20);
    final private static Fraction BONUS_FRACTION_MAX = new Fraction(0.25);
    final private static int CLONES_PER_COMBO_MAX = 15;
    final private static int HAND_SIZE_MAX = 20;
    final public static int HANDS_DEALT_MAX = 10;

    // embedded classes

    private static class AttrRow
        extends Box
    {
        final private JSpinner spinner;
        final private SpinnerNumberModel model;

        public AttrRow(ParmBox3 parent, int iAttr) {
            super(BoxLayout.LINE_AXIS);

            model = new SpinnerNumberModel(
                    Attr.COUNT_DEFAULT,
                    Attr.COUNT_MIN,
                    Attr.COUNT_MAX,
                    1);
            spinner = new JSpinner(model);
            spinner.setEditor(new JSpinner.NumberEditor(spinner, "#"));
            spinner.addChangeListener(parent);
            spinner.setMaximumSize(SMALL);

            final String text = String.format("The %s attribute takes ",
                    StringExt.ordinal(1 + iAttr));

            setAlignmentX(Box.LEFT_ALIGNMENT);
            add(Box.createHorizontalStrut(30));
            add(new JLabel(text));
            add(spinner);
            add(new JLabel(" possible values."));
        }

        public int getValue() {
            return model.getNumber().intValue();
        }

        public void setValue(int attrCount) {
            model.setValue(attrCount);
        }
    }

    // per-instance fields, sorted by type
    private GameOpt gameOpt = null;
    private HandOpt[] handOpts = null;
    final private AttrRow[] attrRows = new AttrRow[GameOpt.ATTR_COUNT_MAX];
    final private JLabel clonesPerComboLabel = new JLabel();
    final private JLabel handsLabel = new JLabel();
    final private JLabel tilesLabel = new JLabel();
    final private JLabel totalCombosLabel = new JLabel();
    final private JLabel totalTilesLabel = new JLabel();
    final private SpinnerNumberModel attrCountModel;
    final private SpinnerNumberModel clonesModel;
    final private SpinnerNumberModel handSizeModel;
    final private SpinnerNumberModel handsDealtModel;
    final private SpinnerNumberModel percentModel;

    // constructors

    @SuppressWarnings("LeakingThisInConstructor")
    public ParmBox3(Wizard wizard) {
        super(wizard, ParmBox3.class.getName());

        // number of hands

        handsDealtModel = new SpinnerNumberModel(
                GameOpt.HANDS_DEALT_DEFAULT,
                GameOpt.HANDS_DEALT_MIN,
                HANDS_DEALT_MAX,
                1);
        final JSpinner handsSpinner = new JSpinner(handsDealtModel);
        handsSpinner.setEditor(new JSpinner.NumberEditor(handsSpinner, "#"));
        handsSpinner.addChangeListener(this);
        handsSpinner.setMaximumSize(SMALL);

        // tiles per hand

        handSizeModel = new SpinnerNumberModel(
                GameOpt.HAND_SIZE_DEFAULT,
                GameOpt.HAND_SIZE_MIN,
                HAND_SIZE_MAX,
                1);
        final JSpinner tilesSpinner = new JSpinner(handSizeModel);
        tilesSpinner.setEditor(new JSpinner.NumberEditor(tilesSpinner, "#"));
        tilesSpinner.addChangeListener(this);
        tilesSpinner.setMaximumSize(SMALL);

        final Box handsRow = new Box(BoxLayout.LINE_AXIS);
        handsRow.add(new JLabel("Deal "));
        handsRow.setAlignmentX(Box.LEFT_ALIGNMENT);
        handsRow.add(handsSpinner);
        handsRow.add(handsLabel);
        handsRow.add(tilesSpinner);
        handsRow.add(tilesLabel);

        // attributes per tile

        attrCountModel = new SpinnerNumberModel(
                GameOpt.ATTR_COUNT_DEFAULT,
                GameOpt.ATTR_COUNT_MIN,
                GameOpt.ATTR_COUNT_MAX,
                1);
        final JSpinner attrCountSpinner = new JSpinner(attrCountModel);
        attrCountSpinner.setEditor(new JSpinner.NumberEditor(
                attrCountSpinner, "#"));
        attrCountSpinner.addChangeListener(this);
        attrCountSpinner.setMaximumSize(SMALL);

        final Box attrCountRow = new Box(BoxLayout.LINE_AXIS);
        attrCountRow.setAlignmentX(Box.LEFT_ALIGNMENT);
        attrCountRow.add(new JLabel("Tiles have "));
        attrCountRow.add(attrCountSpinner);
        attrCountRow.add(new JLabel(" attributes:"));

        // values per attribute

        for (int iAttr = 0; iAttr < GameOpt.ATTR_COUNT_MAX; iAttr++) {
            attrRows[iAttr] = new AttrRow(this, iAttr);
        }

        final Box totalCombosRow = new Box(BoxLayout.LINE_AXIS);
        totalCombosRow.setAlignmentX(Box.LEFT_ALIGNMENT);
        totalCombosRow.add(Box.createHorizontalStrut(60));
        totalCombosRow.add(totalCombosLabel);

        // clones per combo

        clonesModel = new SpinnerNumberModel(
                GameOpt.CLONES_PER_COMBO_DEFAULT,
                GameOpt.CLONES_PER_COMBO_MIN,
                CLONES_PER_COMBO_MAX,
                1);
        final JSpinner clonesSpinner = new JSpinner(clonesModel);
        clonesSpinner.setEditor(new JSpinner.NumberEditor(clonesSpinner, "#"));
        clonesSpinner.addChangeListener(this);
        clonesSpinner.setMaximumSize(SMALL);

        final Box clonesRow = new Box(BoxLayout.LINE_AXIS);
        clonesRow.setAlignmentX(Box.LEFT_ALIGNMENT);
        clonesRow.add(new JLabel("Add "));
        clonesRow.add(clonesSpinner);
        clonesRow.add(clonesPerComboLabel);

        final Box totalTilesRow = new Box(BoxLayout.LINE_AXIS);
        totalTilesRow.setAlignmentX(Box.LEFT_ALIGNMENT);
        totalTilesRow.add(Box.createHorizontalStrut(60));
        totalTilesRow.add(totalTilesLabel);

        // percent bonus tiles

        percentModel = new SpinnerNumberModel(
                GameOpt.BONUS_FRACTION_DEFAULT.toPercent(),
                GameOpt.BONUS_FRACTION_MIN.toPercent(),
                BONUS_FRACTION_MAX.toPercent(),
                1.0);
        final JSpinner percentSpinner = new JSpinner(percentModel);
        percentSpinner.setEditor(new JSpinner.NumberEditor(percentSpinner, "#"));
        percentSpinner.addChangeListener(this);
        percentSpinner.setMaximumSize(SMALL);

        final Box percentRow = new Box(BoxLayout.LINE_AXIS);
        percentRow.setAlignmentX(Box.LEFT_ALIGNMENT);
        percentRow.add(percentSpinner);
        percentRow.add(new JLabel("% of tiles have bonus value."));

        // combine into a single Box

        final Box mainBox = new Box(BoxLayout.PAGE_AXIS);
        mainBox.add(handsRow);
        mainBox.add(Box.createVerticalStrut(8));
        mainBox.add(attrCountRow);
        for (int iAttr = 0; iAttr < GameOpt.ATTR_COUNT_MAX; iAttr++) {
            mainBox.add(attrRows[iAttr]);
        }
        mainBox.add(totalCombosRow);
        mainBox.add(Box.createVerticalStrut(8));
        mainBox.add(clonesRow);
        mainBox.add(totalTilesRow);
        mainBox.add(Box.createVerticalStrut(8));
        mainBox.add(percentRow);

        add(mainBox);
    }

    // methods, sorted by name

    @Override
    public void backAction() {
        updateModel();
        showCard(ParmBox2.class.getName(), gameOpt, handOpts);
    }


    @Override
    public String getTitle() {
        return "Hand and Tile Parameters";
    }

    @Override
    public void nextAction() {
        updateModel();
        showCard(HandBox.class.getName(), gameOpt, handOpts);
    }

    @Override
    public void setModels(Object[] models) {
        assert models != null;
        assert models.length == 2 : models.length;
        assert models[0] != null;
        assert models[0] instanceof GameOpt;
        assert models[1] != null;
        assert models[1] instanceof HandOpt[];

        gameOpt = (GameOpt)models[0];
        gameOpt.validate();
        handOpts = (HandOpt[])models[1];

        updateView();
    }

    @Override
    public void stateChanged(javax.swing.event.ChangeEvent event) {
        // from a spinner
        updateModel();
    }

    private void updateModel() {
        // read spinner values

        final Number handsDealt = (Number)handsDealtModel.getValue();
        gameOpt.setHandsDealt(handsDealt.intValue());

        final Number handSize = (Number)handSizeModel.getValue();
        gameOpt.setHandSize(handSize.intValue());

        final Number attrCount = (Number)attrCountModel.getValue();
        gameOpt.setAttrCount(attrCount.intValue());

        for (int iAttr = 0; iAttr < GameOpt.ATTR_COUNT_MAX; iAttr++) {
            final int valueCount = attrRows[iAttr].getValue();
            gameOpt.setAttrValueCount(iAttr, valueCount);
        }

        final Number clonesPerCombo = (Number)clonesModel.getValue();
        gameOpt.setClonesPerCombo(clonesPerCombo.intValue());

        final Number percent = (Number)percentModel.getValue();
        gameOpt.setBonusFraction(new Fraction(percent.doubleValue() / 100.0));

        gameOpt.validate();

        updateView();
    }

    private void updateView() {
        // the wizard's back/next buttons
        wizard.getBackButton().setText("\u21d0 Back");
        wizard.getBackButton().setEnabled(true);
        wizard.getBackButton().setVisible(true);
        wizard.getNextButton().setText("Next \u21d2");
        wizard.getNextButton().setEnabled(true);

        // set labels, spinner values, and visibility
        final int handsDealt = gameOpt.getHandsDealt();
        handsDealtModel.setValue(handsDealt);
        final String handsText = String.format(" hand%s of ",
                StringExt.plural(handsDealt));
        handsLabel.setText(handsText);

        final int handSize = gameOpt.getHandSize();
        handSizeModel.setValue(handSize);
        final String tilesText = String.format(" tile%s%s.",
                StringExt.plural(handSize),
                handsDealt > 1 ? " each" : "");
        tilesLabel.setText(tilesText);

        final int attrCount = gameOpt.getAttrCount();
        attrCountModel.setValue(attrCount);
        for (int iAttr = 0; iAttr < GameOpt.ATTR_COUNT_MAX; iAttr++) {
            final int valueCount = gameOpt.getAttrValueCount(iAttr);
            attrRows[iAttr].setValue(valueCount);
            attrRows[iAttr].setVisible(iAttr < attrCount);
        }

        final String totalCombosText = String.format(
                "\u21d2 %d attribute combinations",
                gameOpt.countCombos());
        totalCombosLabel.setText(totalCombosText);

        final int clonesPerCombo = gameOpt.getClonesPerCombo();
        clonesModel.setValue(clonesPerCombo);

        final String clonesText = String.format(" clone%s of each combination",
                StringExt.plural(clonesPerCombo));
        clonesPerComboLabel.setText(clonesText);

        final String totalTilesText = String.format("\u21d2 %d tiles in total",
                gameOpt.countTiles());
        totalTilesLabel.setText(totalTilesText);

        percentModel.setValue(gameOpt.getBonusFraction().toPercent());
    }
}
